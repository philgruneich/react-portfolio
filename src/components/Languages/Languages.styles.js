import styled from 'styled-components';
import { pxToEm, pxToRem } from '~/utils/cssMixins';

export const LanguagesContainer = styled.div`
  color: #fff;
  display: grid;
  grid-auto-flow: column;
  grid-gap: ${pxToRem(8)};
`;

export const LanguagesButton = styled.button.attrs({
  type: 'button'
})`
  appearance: none;
  margin: 0;
  padding: 0;
  border-width: 0 0 2px;
  border-style: solid;
  border-color: transparent;
  border-radius: 0;
  background: none;
  color: #fff;
  font-size: ${pxToEm(16)};
  opacity: .666;
  cursor: pointer;

  ${props => props.hasHighContrast && `
    opacity: 1;
  `}

  ${props => props.isActive && `
    border-bottom-color: currentColor;
    opacity: 1;
  `}
`;
