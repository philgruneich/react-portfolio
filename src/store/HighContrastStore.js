import React, { useState, createContext } from 'react';

export const HighContrastContext = createContext();
HighContrastContext.displayName = 'HighContrastContext';

export const HighContrastStore = props => {
  const [hasContrast, saveHasContrast] = useState(false);

  const toggleContrast = () => {
    saveHasContrast(() => !hasContrast);
  }

  return (
    <HighContrastContext.Provider
      value={{
        state: hasContrast,
        dispatch: saveHasContrast,
        toggle: toggleContrast
      }}
      {...props}
    />
  )
}
