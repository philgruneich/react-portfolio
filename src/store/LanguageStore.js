import React, { useState, useEffect, createContext } from 'react';
import { languages } from '~/lang';

export const LanguageContext = createContext();
LanguageContext.displayName = 'LanguageContext';

export const LanguageStore = props => {
  const [lang, saveLang] = useState('en');
  const [translations, saveTranslations] = useState(languages[lang]);

  useEffect(() => {
    saveTranslations(languages[lang]);
  }, [lang]);

  return (
    <LanguageContext.Provider
      value={{
        state: { lang, translations },
        dispatch: saveLang
      }}
      {...props}
    />
  )
}
