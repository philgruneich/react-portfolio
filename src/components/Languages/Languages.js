import React, { useContext } from 'react';
import { LanguageContext } from '~/store/LanguageStore';
import { HighContrastContext } from '~/store/HighContrastStore';
import {
  LanguagesContainer,
  LanguagesButton
} from './Languages.styles';

export const Languages = () => {
  const langContext = useContext(LanguageContext);
  const contrastContext = useContext(HighContrastContext);

  const onClickHandler = langName => {
    langContext.dispatch(langName);
  }

  const langButtons = Langs.map((
    { shortName, langName, title }
  ) => (
    <LanguagesButton
      key={shortName}
      aria-pressed={langContext.state.lang === langName}
      aria-label={title}
      isActive={langContext.state.lang === langName}
      hasHighContrast={contrastContext.state}
      onClick={() => onClickHandler(langName)}
    >{shortName}</LanguagesButton>
  ))

  return (
    <LanguagesContainer aria-label="Switch to another language">
      {langButtons}
    </LanguagesContainer>
  )
};

const Langs = [
  {
    shortName: 'EN',
    langName: 'en',
    title: 'Translate to English'
  },
  {
    shortName: 'PT',
    langName: 'pt-BR',
    title: 'Traduzir para PortuguÃªs'
  }
];
