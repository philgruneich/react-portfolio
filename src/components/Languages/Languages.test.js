import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { LanguageStore } from '~/store/LanguageStore';
import { HighContrastStore } from '~/store/HighContrastStore';
import { Languages } from './Languages';

const defaultLanguageState = {
  state: {
    lang: 'en'
  },
  dispatch: jest.fn()
};

const defaultHighContrastState = {
  state: false
};

test('Matches snapshot', () => {
  const { asFragment } = render(
    <LanguageStore value={defaultLanguageState}>
      <HighContrastStore value={defaultHighContrastState}>
        <Languages />
      </HighContrastStore>
    </LanguageStore>
  );

  expect(asFragment()).toMatchSnapshot();
});

test('Matches snapshot with high contrast', () => {
  const { asFragment } = render(
    <LanguageStore value={defaultLanguageState}>
      <HighContrastStore value={{
        state: true
      }}>
        <Languages />
      </HighContrastStore>
    </LanguageStore>
  );

  expect(asFragment()).toMatchSnapshot();
});

test('Clicking button toggles high contrast', () => {
  const { getByLabelText } = render(
    <LanguageStore value={defaultLanguageState}>
      <HighContrastStore value={defaultHighContrastState}>
        <Languages />
      </HighContrastStore>
    </LanguageStore>
  );

  const button = getByLabelText('Traduzir para PortuguÃªs');

  fireEvent.click(button);

  expect(defaultLanguageState.dispatch).toBeCalled();
});
