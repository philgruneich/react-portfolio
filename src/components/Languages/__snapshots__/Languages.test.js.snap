// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
<DocumentFragment>
  <div
    aria-label="Switch to another language"
    class="Languagesstyles__LanguagesContainer-sc-1i72sgo-0 dxNjpm"
  >
    <button
      aria-label="Translate to English"
      aria-pressed="true"
      class="Languagesstyles__LanguagesButton-sc-1i72sgo-1 KIPAK"
      type="button"
    >
      EN
    </button>
    <button
      aria-label="Traduzir para Português"
      aria-pressed="false"
      class="Languagesstyles__LanguagesButton-sc-1i72sgo-1 gHsmBy"
      type="button"
    >
      PT
    </button>
  </div>
</DocumentFragment>
`;

exports[`Matches snapshot with high contrast 1`] = `
<DocumentFragment>
  <div
    aria-label="Switch to another language"
    class="Languagesstyles__LanguagesContainer-sc-1i72sgo-0 dxNjpm"
  >
    <button
      aria-label="Translate to English"
      aria-pressed="true"
      class="Languagesstyles__LanguagesButton-sc-1i72sgo-1 gKRrif"
      type="button"
    >
      EN
    </button>
    <button
      aria-label="Traduzir para Português"
      aria-pressed="false"
      class="Languagesstyles__LanguagesButton-sc-1i72sgo-1 eqBguT"
      type="button"
    >
      PT
    </button>
  </div>
</DocumentFragment>
`;
